### PR TITLE
wavm: emit stack trace when an exception happens

### PR DESCRIPTION
**Pull request by mathetake**
this PR emits call stack trace when exception happens in WAVM just like V8 (#66)

this requires a change in the WAVM library used in Envoy: https://github.com/envoyproxy/envoy/pull/13792

```
[2020-10-28 14:24:10.551][979699][trace][wasm] [source/extensions/common/wasm/wasm_vm.cc:29] Function proxy_on_tick failed:
WAVM message: wavm.reachedUnreachable
wasm backrace:
  1: wasm!helloworld!runtime._panic+7
  2: wasm!helloworld!command-line-arguments.C+2
  3: wasm!helloworld!command-line-arguments.B+0
  4: wasm!helloworld!command-line-arguments.A+0
  5: wasm!helloworld!(*command-line-arguments.helloWorld).OnTick+74
  6: wasm!helloworld!proxy_on_tick+0
```

Signed-off-by: mathetake <takeshi@tetrate.io>

### CODE REVIEW

File: src/wavm/wavm.cc
Comment (mathetake) [inline src/wavm/wavm.cc:108]: as a comparison, the following is an example of full call stack messages
```
[2020-10-28 13:12:44.953][785800][trace][wasm] [source/extensions/common/wasm/wasm_vm.cc:29] Function: proxy_on_tick failed: wavm.reachedUnreachable
Call stack:
  host!envoy+112911829
  wasm!helloworld!runtime._panic+7
  wasm!helloworld!command-line-arguments.C+2
  wasm!helloworld!command-line-arguments.B+0
  wasm!helloworld!command-line-arguments.A+0
  wasm!helloworld!(*command-line-arguments.helloWorld).OnTick+74
  wasm!helloworld!proxy_on_tick+0
  thnk!C to WASM thunk!i32->()+0
  host!envoy+112252692
  host!envoy+112252572
  host!envoy+112252476
  host!envoy+112252252
  host!envoy+85651453
  host!envoy+111949395
  host!envoy+111949348
  host!envoy+112066521
  host!envoy+111948228
  host!envoy+112251839
  host!envoy+109799095
  host!envoy+109798956
  host!envoy+109798876
  host!envoy+109798556
  <5 redundant frames omitted>
  host!envoy+111948024
  host!envoy+109797831
  host!envoy+109797585
  host!envoy+109797462
 ```

**Comment by mathetake** [comment]
blocker: https://github.com/proxy-wasm/proxy-wasm-cpp-host/issues/49

We couldn't tell even if this compile. Before we merge this PR, I will write WAVM tests, and then test this PR

File: src/wavm/wavm.cc
Comment (PiotrSikora) [inline src/wavm/wavm.cc:103]: Could you pull the message into the first line and remove `WAVM message:`?

Also, maybe switch the `wasm backtrace:` to `Proxy-Wasm plugin in-VM backtrace:` to make it clear that it covers only the plugin part and not host.

The example message should look like this:
```
[2020-10-28 14:24:10.551][979699][trace][wasm] [source/extensions/common/wasm/wasm_vm.cc:29] Function proxy_on_tick failed: wavm.reachedUnreachable
Proxy-Wasm plugin in-VM backtrace:
  1: wasm!helloworld!runtime._panic+7
  2: wasm!helloworld!command-line-arguments.C+2
  3: wasm!helloworld!command-line-arguments.B+0
  4: wasm!helloworld!command-line-arguments.A+0
  5: wasm!helloworld!(*command-line-arguments.helloWorld).OnTick+74
  6: wasm!helloworld!proxy_on_tick+0
```

**Review by PiotrSikora — CHANGES_REQUESTED** [review]
I don't think we need to block this on tests in this repo if you can open a PR in Envoy to make sure that this doesn't break anything.

File: src/wavm/wavm.cc
Comment (mathetake) [reply src/wavm/wavm.cc]: fixed 
Screenshot from 2020-10-29 18-36-57

File: src/wavm/wavm.cc
Comment (PiotrSikora) [inline src/wavm/wavm.cc:116]: Actually, one more thing. Could you add leading spaces here, so that single-digit and double-digit numbers have the same width? Right now, it's going to look a bit ugly once we hit 10:
```
[2020-10-28 14:24:10.551][979699][trace][wasm] [source/extensions/common/wasm/wasm_vm.cc:29] Function proxy_on_tick failed: wavm.reachedUnreachable
Proxy-Wasm plugin in-VM backtrace:
  1: wasm!helloworld!runtime._panic+7
  2: wasm!helloworld!command-line-arguments.C+2
  3: wasm!helloworld!command-line-arguments.B+0
  4: wasm!helloworld!command-line-arguments.A+0
  5: wasm!helloworld!(*command-line-arguments.helloWorld).OnTick+74
  6: wasm!helloworld!proxy_on_tick+0
  7: wasm!helloworld!proxy_on_tick+0
  8: wasm!helloworld!proxy_on_tick+0
  9: wasm!helloworld!proxy_on_tick+0
  10: wasm!helloworld!proxy_on_tick+0
```

I think you could use `std::ostringstream` with `<< std::setfill(' ') << std::setw(3)`, I'm not sure if there is any easier way to do that in C++.

**Review by PiotrSikora — CHANGES_REQUESTED** [review]

File: src/wavm/wavm.cc
Comment (mathetake) [reply src/wavm/wavm.cc:116]: yeah, that's true. I'm on it

Comment (mathetake) [reply src/wavm/wavm.cc]: done

Screenshot from 2020-10-29 19-48-34

**Review by PiotrSikora — APPROVED** [review]

**Pull request merged**
